hail/python/hailtop/config/deploy_config.py

```diff
@@ -133,6 +135,58 @@ async def get_metrics(_):
         log.info(f'serving paths at {base_path}')
         return root_app
 
+    def client_ssl_context(self) -> ssl.SSLContext:
+        if self._location == 'k8s':
+            return internal_client_ssl_context()
+        # no encryption on the internal gateway
+        return external_client_ssl_context()
+
+    def server_ssl_context(self) -> Optional[ssl.SSLContext]:
+        if self._location == 'k8s':
+            return internal_server_ssl_context()
+        # local mode does not have access to self-signed certs
+        return None
+
+
+class TerraDeployConfig(DeployConfig):
```

Comment:
It seems to me that a namespace is fully identified by its domain & subpath. If you replace the explicit namespace name with a pair of domain and subpath everywhere it's used, would that work? You'll need to migrate the tokens files or just force everyone to log in again so that the pair is used as the key. Instead of saying "Logged in to namespace XXX" you can just say "Logged in to domain/subpath". I think your suggestion in the main thread comment works: if subpath is empty, you put the service in the domain name otherwise you append it.

Reply:
I'll try this out. I'm slightly worried about backward compatibility but I think that's only a concern for dev namespaces, which I don't care about being backwards compatible with old clients. For production users, the only change should be the added field of an empty subpath.

Reply:
I don't think we can immediately remove `default_namespace`, old clients still expect it to exist. But we can basically stop using it in new code.